tech/post/2017-09-25-Kubernetes-Scheduler.md

```diff
@@ -0,0 +1,145 @@
+
+![](../images/2017-09-25-Kubernetes-Scheduler/Scheduler.png)
+
+## 2. Scheduler Workflow
+
+The Kubernetes Scheduler is focused on scheduling pods into specified work nodes by schedule algorithm, which also be called ```Bind```.The input of Scheduler is the pods which need to be scheduled and the nodes which could be scheduled into, and the output of Scheduler is ```Bind``` which connects between pods and nodes.
+
+![](../images/2017-09-25-Kubernetes-Scheduler/Workflow.png)
+
+The Kubernetes Scheduler tries to find a node for each Pod, one at a time.
+- First it applies a set of "predicates" to filter out inappropriate nodes. For example, if the PodSpec specifies resource requests, then the Scheduler will filter out nodes that don't have at least that much resources available.
+- Second, it applies a set of "priority functions" that rank the nodes that weren't filtered out by the predicate check. For example, it tries to spread Pods across nodes and zones while at the same time favoring the least loaded nodes.
+- Finally, the node with the highest priority is chosen, if there are multiple such nodes, then one of them is chosen at random.
+
+![](../images/2017-09-25-Kubernetes-Scheduler/Algorithm.png)
+
+## 3. Scheduler Code Analysis
+
+Scheduler main implementations include ```plugin/cmd/kube-scheduler``` and ```plugin/pkg/scheduler```.
+- ```plugin/cmd/kube-scheduler``` starts the Scheduler process and reads the configurations from Command Line Interface and so on.
+- ```plugin/pkg/scheduler/algorithm``` includes two kinds of scheduling algorithms: ```predicates``` and ```priorities```.
+- ```plugin/pkg/scheduler/algorithmprovider``` is used to register and configure Scheduler algorithm providers.
+- ```plugin/pkg/scheduler/factory``` creates scheduling algorithm and scheduling parameters.
+- ```plugin/pkg/scheduler/scheduler.go``` is the main class of Scheduler.
+- ```plugin/pkg/scheduler/core/generic_scheduler.go``` invokes the detail scheduling algorithm.
+
+The entrance of Scheduler is ```plugin/cmd/kube-scheduler/scheduler.go```.
+- First Scheduler creates ```configFactory``` and then creates ```config```.
+- If it specifies policy-config-file, Scheduler will create ```config``` by policy-config-file(CreateFromConfig),
+- If not, Scheduler will create ```config``` by Scheduler algorithm providers(CreateFromProvider),
+- After Scheduler have got all the parameters, the scheduling work will be executed in ```plugin/pkg/scheduler/scheduler.go```.
+- Scheduler invokes ```scheduleOne``` function and ```getNextPod``` from pod queue.
+- Then Scheduler execute ```schedule``` function, and actually the schedule process will be executed in ```plugin/pkg/scheduler/core/generic_scheduler.go```.
+- If the operation is successful, the Scheduler will get a most fit node for this pod.

```

Comment:
S/a most fit/the fittest

Reply:
I guess 'most fit' is correct, Thanks your review.